### PR TITLE
feat: print functionality [RWDQA-1]

### PR DESCRIPTION
**Pull request by KaiVandivier**
https://dhis2.atlassian.net/browse/RWDQA-1

Getting the styles right is pretty finicky, but here's a good start imo

To do:
- [x] Check on if we want a "Print" button or just use the browser function
- [x] Avoid blank first page in Firefox
- [x] Make sure graph layouts reflow to the correct size before printing (they can get "stuck" in their original size if the window size is very different from the print page size, making the chart look too small or too large, potentially extending off the page)

Done:
1. Fix overflow to be able to print more than just the visible viewport
2. Hide the DHIS2 header bar and the report selector bar
3. Add & avoid page breaks where appropriate
4. Scale the page to an appropriate size for print

Example PDF:
[Data Quality Annual Report _ DHIS2.pdf](https://github.com/hisprwanda/who-data-quality-annual-report/files/13481109/Data.Quality.Annual.Report._.DHIS2.pdf)

**Update - again**

I was able to use Highcharts to reflow the charts on print so they are the right size on the print page in the end 👍 

Here are some examples of the chart size errors which have been fixed (notice some graphs are truncated):
chart-size-2
chart-size-1

A known edge case is that a section header can end up on a different page than the rest of its section. This is also minor and depends on the report contents. The best solution would be to find a way to group the section header with the first section item and apply the `break-inside: avoid;` CSS property to that group, but that requires some new HTML structure and some logic to handle the first section item differently from the others and also account for an empty section. Ticket here: [RWDQA-98](https://dhis2.atlassian.net/browse/RWDQA-98)

Example:
orphaned-heading

### CODE REVIEW

File: src/App.module.css [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
-    /* font-size: 1rem; */
+    height: 100%;
+}
+
+@media print {
+    /* a bit of a hack to get the sizes right during printing (assuming
+    this is printed on A4/US Letter paper): 
+    https://stackoverflow.com/a/66593515 */
+    @page {
+        size: A3;
```
Comment (tomzemp) [inline src/App.module.css:15]: 😆 Hah. This is is like the opposite of what I would expect (setting it to a larger size than what you're printing to fixes the scaling issues 🤪)

**Review by tomzemp — APPROVED** [review]
Looks great! 🤩

Two thoughts:

- the background-color on the subheaders is not appearing for me in print. If I add `print-color-adjust: exact;` to `sectionHeading` class it fixes the issue (not sure if you have other ideas, or if it was intentional to not get this background).

- I think it's a good idea to hide the report parameter selector bar in print, but it would be nice to have the context so that people know what the report is for. We could either add the selections in a section that we hide in print, or we could just hide the buttons of the report bar.

**Comment by KaiVandivier** [comment]
Cool, I think it's a good idea to "force" the color to the section headers in print -- if the user doesn't have "show backgrounds" selected on print, they look kinda bad (I think the rest of the report looks okay though without backgrounds), and the user can always choose to print grayscale

I'll also see about adding in some indication of the selected context in print, and add the button that says "print"

**Comment by KaiVandivier** [comment]
Summary of current changes:
context-and-header-color

Currently, Highcharts does not size correctly using `window.print()` -- here's a [Highcharts thread](https://github.com/highcharts/highcharts/issues/16931), and a seemingly related [DOMPurify thread](https://github.com/cure53/DOMPurify/issues/529). I haven't found another way to print from Javascript, so another resizing solution is needed. It might be possible to do resizing work in JS before calling `window.print()`, for example

File: src/pages/AnnualReport.js [hunk 2]
```diff
@@ -7,21 +7,47 @@ import styles from './AnnualReport.module.css'
...
+     * function. 1 sec seems long enough on this 2020 Macbook Pro with 6x
+     * CPU slowdown using Chrome dev tools
+     */
+    const print = useCallback(() => {
+        setPrinting(true)
+        setTimeout(() => {
+            window.print()
+            setPrinting(false)
+        }, 1000)
+    }, [])
```
Comment (KaiVandivier) [inline src/pages/AnnualReport.js:32]: I made this workaround which seems to be smooth and effective 👍

File: src/components/annual-report/ReportPrintHeader.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            <h1 className={styles.title}>
+                {i18n.t('Data Quality Annual Report')}
+            </h1>
+            <div className={styles.parametersContainer}>
+                <p className={styles.parameterLine}>
+                    {/**
+                     * Reminder: ':' is the namespace separater for i18next so
+                     * this needs interpolation:
+                     */}
+                    <strong>{i18n.t('Data{{c}} ', { c: ':' })}</strong>
```
Comment (tomzemp) [inline src/components/annual-report/ReportPrintHeader.js:42]: you can also override the separator in the options you pass to the translation function, e.g.
https://github.com/hisprwanda/who-data-quality-annual-report/blob/main/src/components/annual-report/common/Warnings.js#L73-L79
docs: https://www.i18next.com/overview/configuration-options

File: src/components/annual-report/ReportPrintHeader.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (!reportParameters || Object.keys(reportParameters).length === 0) {
+        return null
+    }
+
+    const { orgUnitName, orgUnitLevelName, periods } = reportParameters
+
+    return (
+        <div className={styles.showOnlyInPrint}>
+            <h1 className={styles.title}>
```
Comment (tomzemp) [inline src/components/annual-report/ReportPrintHeader.js:33]: Semantically, I guess this makes sense, but feels a bit strange that our h1 tag is generally hidden?

**Review by tomzemp — APPROVED** [review]
looks good! 🖨

File: src/components/annual-report/ReportPrintHeader.js (hunk 3 above)
Comment (KaiVandivier) [reply src/components/annual-report/ReportPrintHeader.js:42]: Ah nice, thank you! I'll use that 🙌

File: src/components/annual-report/ReportPrintHeader.js (hunk 4 above)
Comment (KaiVandivier) [reply src/components/annual-report/ReportPrintHeader.js:33]: Heh, yeah maybe... it's the only one in the app though, and it makes sense in print, so it seems good enough to me though 🙂

**Pull request merged**